File: backend/SpringBoot_Service/src/main/java/com/postgresql/SpringBoot_Service/ExamController.java
package com.postgresql.SpringBoot_Service;

import com.postgresql.SpringBoot_Service.model.*;
import com.postgresql.SpringBoot_Service.repo.*;
import com.postgresql.SpringBoot_Service.util.GradeCalculator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@CrossOrigin(origins = "http://localhost:5173")
@RestController
@RequestMapping("/api/exams")
public class ExamController {

    @Autowired
    private ExamRepo examRepo;

    @Autowired
    private FacultyStudentRepo facultyStudentRepo;

    @Autowired
    private StudentExamRegistrationRepo registrationRepo;

    @Autowired
    private FacultyProfessorRepo professorRepo;

    // Get all exams
    @GetMapping
    public ResponseEntity<List<FacultyExam>> getAllExams() {
        return ResponseEntity.ok(examRepo.findAll());
    }

    // Create new exam (professor only)
    @PostMapping
    public ResponseEntity<?> createExam(@RequestBody FacultyExam exam) {
        try {
            return ResponseEntity.ok(examRepo.save(exam));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    // Register student for exam
    @PostMapping("/{examId}/register")
    public ResponseEntity<?> registerForExam(
            @PathVariable Long examId,
            @RequestBody Map<String, Long> request) {
        Long studentId = request.get("studentId");
        
        FacultyExam exam = examRepo.findById(examId)
            .orElseThrow(() -> new RuntimeException("Exam not found"));

        FacultyStudent student = facultyStudentRepo.findById(studentId)
            .orElseThrow(() -> new RuntimeException("Student not found"));

        if (registrationRepo.existsByStudentAndExam(student, exam)) {
            return ResponseEntity.badRequest()
                .body("Student already registered for this exam");
        }

        StudentExamRegistration registration = new StudentExamRegistration();
        registration.setStudent(student);
        registration.setExam(exam);
        registration.setRegistrationDate(LocalDateTime.now());
        registration.setStatus(StudentExamRegistration.RegistrationStatus.REGISTERED);

        return ResponseEntity.ok(registrationRepo.save(registration));
    }

    // Get student's registered exams
    @GetMapping("/student/{studentId}")
    public ResponseEntity<List<StudentExamRegistration>> getStudentExams(
            @PathVariable Long studentId) {
        FacultyStudent student = facultyStudentRepo.findById(studentId)
            .orElseThrow(() -> new RuntimeException("Student not found"));
        return ResponseEntity.ok(registrationRepo.findByStudent(student));
    }

    // Get professor's exams
    @GetMapping("/professor/{professorId}")
    public ResponseEntity<List<FacultyExam>> getProfessorExams(@PathVariable Long professorId) {
        FacultyProfessor professor = professorRepo.findById(professorId)
            .orElseThrow(() -> new RuntimeException("Professor not found"));
        return ResponseEntity.ok(examRepo.findByProfessor(professor));
    }

    // Update exam grade
    @PutMapping("/{registrationId}/grade")
    public ResponseEntity<?> updateGrade(
            @PathVariable Long registrationId,
            @RequestBody Map<String, Integer> request) {
        try {
            Integer points = request.get("points");
            if (points == null || points < 0 || points > 100) {
                return ResponseEntity.badRequest()
                    .body(Map.of("error", "Points must be between 0 and 100"));
            }
            
            StudentExamRegistration registration = registrationRepo.findById(registrationId)
                .orElseThrow(() -> new RuntimeException("Registration not found"));
                
            // Calculate grade from points
            int grade = GradeCalculator.calculateGrade(points);
            boolean isCompleted = GradeCalculator.isSubjectCompleted(grade);
            String gradeDescription = GradeCalculator.getGradeDescription(points);
                
            registration.setPoints(points);
            registration.setGrade(grade);
            registration.setStatus(isCompleted ? 
                StudentExamRegistration.RegistrationStatus.PASSED : 
                StudentExamRegistration.RegistrationStatus.FAILED);
                
            registration = registrationRepo.save(registration);
            
            Map<String, Object> response = new HashMap<>();
            response.put("points", points);
            response.put("grade", grade);
            response.put("status", registration.getStatus());
            response.put("description", gradeDescription);
            response.put("completed", isCompleted);
            
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.badRequest()
                .body(Map.of("error", e.getMessage()));
        }
    }
}

File: backend/src/controllers/exam.controller.ts
import { Request, Response } from "express";
import { ExamService } from "../services/exam.service";

export class ExamController {
  /**
   * POST /api/exams
   */
  static async createExam(req: Request, res: Response) {
    try {
      const { subjectId, professorId, examTime, location, maxPoints } =
        req.body;

      if (!subjectId || !professorId || !examTime) {
        return res.status(400).json({ error: "Missing required fields" });
      }

      const exam = await ExamService.createExam({
        subjectId: Number(subjectId),
        professorId: Number(professorId),
        examTime,
        location,
        maxPoints,
      });

      return res.json({ message: "Exam created", exam });
    } catch (error: any) {
      console.error("Create exam error:", error);
      return res
        .status(400)
        .json({ error: error.message || "Failed to create exam" });
    }
  }

  /**
   * GET /api/exams/professor/:id
   */
  static async getExamsByProfessor(req: Request, res: Response) {
    try {
      const id = parseInt(req.params.id);
      const exams = await ExamService.getExamsByProfessor(id);
      return res.json(exams);
    } catch (error) {
      console.error("Get exams error:", error);
      return res.status(500).json({ error: "Failed to fetch exams" });
    }
  }

  /**
   * GET /api/exams/student/:email
   */
  static async getExamsByStudent(req: Request, res: Response) {
    try {
      const { email } = req.params;
      const exams = await ExamService.getExamsByStudent(email);

      if (exams === null) {
        return res.status(404).json({ error: "Student not found" });
      }

      return res.json(exams);
    } catch (error) {
      console.error("Get student exams error:", error);
      return res.status(500).json({ error: "Failed to fetch student exams" });
    }
  }

  /**
   * GET /api/exams/:id
   */
  static async getExamById(req: Request, res: Response) {
    try {
      const id = parseInt(req.params.id);
      const exam = await ExamService.getExamById(id);

      if (!exam) {
        return res.status(404).json({ error: "Exam not found" });
      }

      return res.json(exam);
    } catch (error) {
      console.error("Get exam error:", error);
      return res.status(500).json({ error: "Failed to fetch exam" });
    }
  }

  /**
   * POST /api/exams/:id/register
   */
  static async registerForExam(req: Request, res: Response) {
    try {
      const examId = parseInt(req.params.id);
      const { email } = req.body;

      if (!email) {
        return res.status(400).json({ error: "Student email required" });
      }

      const registration = await ExamService.registerForExam(examId, email);
      return res.json({ message: "Registered successfully", registration });
    } catch (error: any) {
      console.error("Register exam error:", error);
      return res
        .status(400)
        .json({ error: error.message || "Failed to register" });
    }
  }

  /**
   * DELETE /api/exams/:id
   */
  static async deleteExam(req: Request, res: Response) {
    try {
      const id = parseInt(req.params.id);
      await ExamService.deleteExam(id);
      return res.json({ message: "Exam deleted successfully" });
    } catch (error: any) {
      console.error("Delete exam error:", error);
      return res
        .status(400)
        .json({ error: error.message || "Failed to delete exam" });
    }
  }
}
